src/directives/repeat.ts

```diff
@@ -41,67 +96,139 @@ export function repeat<T>(
+    // * Iterate old & new lists from both sides, updating, swapping, or
+    // removing
+    //   items at the head/tail locations until neither head nor tail can move
+    // * Once head and tail cannot move, any mismatches are due to either new or
+    //   moved items; create and insert new items, or else move old part to new
+    //   position using the "old key to old index" map.
+    // * If either the new or old pointers move past each other then all we have
+    //   left is additions (if old list exhsusted) or removals (if new list
+    //   exhausted). Those are handled in the final while loops at the end.
+    // * TODO(kschaaf) Note, we could calculate the longest increasing
+    //   subsequence (LIS) of old items in new position, and only move those not
+    //   in the LIS set. However that costs O(nlogn) time and adds a bit more
+    //   code, and only helps make rare types of mutations require fewer moves.
+    //   The above handles removes, adds, reversal, swaps, and single moves of
+    //   contiguous items in linear time, in the minimum number of moves. As
+    //   the number of multiple moves where LIS might help approaches a random
+    //   shuffle, the LIS optimization becomes less helpful, so it seems not
+    //   worth the code at this point, but could consider if a compelling case
+    //   arises.
+    while (oldStartIndex <= oldEndIndex && newStartIndex <= newEndIndex) {
+      if (oldStartPart == null) {
+        // Old part at head has already been used; skip
+        oldStartPart = oldParts[++oldStartIndex];
+      } else if (oldEndPart == null) {
+        // Old part at tail has already been used; skip
+        oldEndPart = oldParts[--oldEndIndex];
+      } else if (oldStartPart.key == newStartResult.key) {
+        // Old head matches new head; update in place
+        newParts[newStartIndex] = updatePart(oldStartPart, newStartResult);
+        oldStartPart = oldParts[++oldStartIndex];
+        newStartResult = newResults[++newStartIndex];
+      } else if (oldEndPart.key == newEndResult.key) {
+        // Old tail matches new tail; update in place
+        newParts[newEndIndex] = updatePart(oldEndPart, newEndResult);
+        oldEndPart = oldParts[--oldEndIndex];
+        newEndResult = newResults[--newEndIndex];
+      } else if (!newKeyToIndexMap.has(oldStartPart.key)) {
+        // Old head is no longer in new list; remove
+        removePart(oldStartPart);
+        oldStartPart = oldParts[++oldStartIndex];
+      } else if (!newKeyToIndexMap.has(oldEndPart.key)) {
+        // Old tail is no longer in new list; remove
+        removePart(oldEndPart);
+        oldEndPart = oldParts[--oldEndIndex];
+      } else if (oldStartPart.key == newEndResult.key) {
+        // Old head matches new tail; update and move to new tail
+        newParts[newEndIndex] = updatePart(oldStartPart, newEndResult);
+        movePart(directivePart, oldStartPart, newParts[newEndIndex + 1]);
```

Comment:
`oldStartPart === newParts[newEndIndex]`. Since the old part is irrelevant, moving the new one would be easier to understand (similar in next condition as well).

Reply:
Hmm.  My mental model was "update and (possibly) move the old part to the new spot", so the operations are all about making old parts correct in the new order.